src/server/services/wms/qgswmsgetlegendgraphics.cpp

```diff
@@ -284,10 +285,11 @@ namespace QgsWms
             legendNode->setUserLabel( QStringLiteral( " " ) );
           }
         }
-        else if ( !parameters.layerTitleAsBool() && model->layerLegendNodes( nodeLayer, true ).count() )
+        else if ( ruleLabel.compare( QStringLiteral( "AUTO" ), Qt::CaseInsensitive ) == 0 )
```

Comment:
shouldn't this be added as a new condition in here ( `( !parameters.layer ... ) || ruleLabel.compare( "AUTO" ... )`?

Reply:
LAYERTITLE is already handled in code, here and it's not needed again.
```
// layer titles - hidden or not
QgsLegendRenderer::setNodeLegendStyle( nodeLayer, parameters.layerTitleAsBool() ? QgsLegendStyle::Subgroup : QgsLegendStyle::Hidden );
```
        

RULELABEL is independent of LAYERTITLE parameter, so it's the case with AUTO value. Logic if labels are visible or not, does, not depend on value of LAYERTITLE parameter. I think this is more clear now.

Reply:
I'm confused. Maybe you can illuminate me. Otherwise I need more time to check it out deeper.

Currently I don't understand when a node is embedded in parent and when it's not. Apparently a single symbol node is embedded and the categories are not.

On the check I did in the mentioned PR, I checked if there are **un**embedded nodes and if so I unembedded the embedded ones (like the single symbol node). I don't really remember everything, but I guess I wanted to avoid that anything gets unembedded, if there is no other child node in there but the single symbol child node.  So when does this happen anyway? What are the embedded nodes beside the single symbol. Are there any at all?

If there are no cases of other embedded nodes, then this unembedding never happens, since there is the possibility of a single symbol (and then there are no unembedded nodes) or of not single symbol (then there might be unembedded nodes but no embedded ones). So there is nothing to unembedd anyway. What would make this clause to unneeded. And with your integration you would reintroduce it for the `RULELABEL=AUTO`, what would be fair enough then.

But if there are other embedded nodes than the single symbol, we need to check what happens to them, when they get unembedded now, and before they would not have been unembedded because it checked `LAYERTITLE=false`.

Reply:
Don't know exactly about embedding nodes, that code was there before and was removing labels for single symbols. I just changed condition when this happens. Before (in 3.4) it happened if LAYERTITLE=FALSE, after your change I believe did not happen at all, so you would have same results just by removing complete `else... `sentence.
Tests are fine, so I really don't see any problems here, specially since we left default behaviour same. Change is only applied when using new value AUTO.

Reply:
Okay, since my last implementation looks weird to me now, I did some more research.

I think, the only point where a nod is embedded is here:
https://github.com/qgis/QGIS/blob/master/src/core/qgsmaplayerlegend.cpp#L338-L339

This means that it embedds a child-node when there is exactly one of 'em, and when there is no label. This is the situation on single symbol or on category / rule with only one section without a label.

So the case of the unembedding never occurs before (after my implementation). So I'm good with your implementation, since it does not change the current behavior.

Reply:
One last thing: When you remove the check for layername, with your `RULELABEL=AUTO` implementation, the label of the child is removed, no matter if `LAYERTITLE` is true or false or unset. When you want to have back the status of 3.4 with `RULELABEL=AUTO`, then it should still check it.

On 3.4: 
- `LAYERTITLE=FALSE` leads to display no Layertitle and no Rulelabel
- `LAYERTITLE=TRUE` or unset leads to display Layertitle and Rulelabel

After your implementation and `RULELABEL=AUTO`: 
- `LAYERTITLE=FALSE` leads to display no Layertitle and no Rulelabel
- `LAYERTITLE=TRUE` or unset leads to display Layertitle and **no** Rulelabel

Is this really what you want? If yes, then fair enough. If not, then leave the `!parameters.layerTitleAsBool() &&` there.

Reply:
I did another test to try categorized legend with 1 label and AUTO parameter. It does not remove labels which is correct, so that embedding part is only for single symbols.

After some discussion in the beginning I removed LAYERTITLE check with RULELABEL=AUTO because I think it's more clear that way. So I'm correcting your part that after my implementation there is no dependency between RULELABEL and LAYERTITLE.

Reply:
The part checking if LAYERTITLE is false, is not mine. It's the one that has been implemented in 3.4. as far as I know. I thought you want to have this behavior back with the AUTO, no?

Reply:
I know this is older, bit I see no need for it now and I think is more clear logic

Reply:
Don't you think it could be possible that an old client relies on that it's only removed when `LAYERTITLE` is false and needs this behavior? I know that I created this regression, but I think if we give em the possibility to have the old behavior back, why introduce another one that is neither like the new nor the old behavior?

But I'll approve the request still. Since it does not make any difference to the master behavior without setting `AUTO` parameter, it's okay.

Reply:
Thanks, only change for old clients to get back used behaviour from 3.4 will be to add fixed AUTO value which is not a problem. This solution is better now, I will also add it into documentation once issue is opened there.

Reply:
So the check for LAYERTITLE has not been in 3.4? Okay, then it's fine. I thought in 3.4 it did display the text of the singlesymbols when LAYERTITLE was true and it did not when LAYERTITLE was false.

Reply:
No, sorry if I wasn't clear, you thought correct. There was a check for LAYERTITLE=false and only in that case labels were removed for single symbols. That's why with this change to get old behaviour client must apply AUTO value. But this is not a problem and like said logic to behave LAYERTITLE and RULELABEL independently is more clear. It does not have to be strict that only case to get RULELABEL=AUTO is when LAYERTITLE is FALSE.